### PR TITLE
[WIP] helm dashboard integration

### PR DESCRIPTION
**Pull request by nicr9**
Hey dashboard guys and gals,

Myself and @joe2far were quite taken with helm's issue for dashboard support: kubernetes/helm#990

We mentioned it to our bosses at [Ammeon](https://github.com/ammeon) and they were similarly excited! In fact they wanted us to mock up a prototype for a recent client demo.

As we were under a tight deadline we hammered together the following changes under the principle of "it only matters if the user will see it". We obviously have a long way to go before it's ready to touch master (the changes currently expect that tiller will always be available) but now that we're no longer operating under a deadline we'd like to start a discussion with yourselves (and the helm guys as well).

We're open to all suggestions for refactoring/improvement, please be kind though - our angular experience is minimal.

Here's some screenshots:

Also, here's a [demo video](https://drive.google.com/open?id=0BwsH9arFx6LBTmx3eDhfX19vWHc)

**Open design questions for interested parties:**
- [ ] Think of a suitable way to store/configure available helm repositories
- [ ] Ability to provide template values for chart at deploy-time
- [ ] How to handle upgrading releases when a newer chart is available?
- [ ] Deleting releases from dashboard
- [ ] What would it mean to edit a release from the dashboard? If this doesn't make sense, remove the option
- [ ] When looking up a specific release's details, how do you specify a namespace? Is it necessary?
- [ ] Learn about ListMeta, ObjectMeta and TypeMeta. Should they be added to the Release types for other dashboard functionality?

**Here's a list of the changes/improvements that we still aim to make:**
- [ ] Link "Releases" in side panel to Releases-specific list page
- [ ] ReleaseList should include failed/broken releases with a coloured icon and warning message in a tooltip
- [ ] Make sure that there's logic for hiding helm related stuff when `helmClient == nil` 
- [ ] Clean up the definitions of types in `externs/backendapi.js`
- [ ] Make sure all annotations are accurate and include a comment
- [ ] More robust error checking
- [ ] Is `ReleaseDetail` type needed?

We've pushed the latest build of our dashboard image to docker.io/ammeon/kubernetes-helm-dashboard-amd64:0.0.13

### CODE REVIEW

**Comment by cheld** [comment]
Wow, cool. I tried to have a look at the changes but my notebook died :-) Question: would it help to squash vendoring and changes into two commits?

**Comment by nicr9** [comment]
Hey @cheld,

Sorry, we didn't realise that would happen! We have recently squashed commits to tidy things up but that was before we saw your comment.

Would it help it we crafted a commit that reversed all the vendoring changes? That way they wouldn't show up in the diff and you could provide review. If/when we decide everything is good to go, we can revert that specific commit and push that in before the merge.

**Comment by joe2far** [comment]
I have now signed the CLA

**Comment by floreks** [comment]
@nicr9 You can try to extract vendoring into separate commit. Once `Files changed` tab is open we can choose commit to review.

In order to review it fully we also need vendor files.

PS. This looks awesome! Thanks for contribution.

**Comment by joe2far** [comment]
Hi @floreks, we have now separated out all vendor updates into a separate commit.
Do you have a advice for fixing the CLA problem? (I haven't had any issues in previous PR's)

**Comment by floreks** [comment]
Ahh... CLA's :) I believe this is related to multiple commits with different emails and bot doesn't know to whom this PR belongs to. I've had similar issue when trying to fix greenkeeper PR by adding my commit. I'm not sure if there is something that we can do to fix that other than single commit with CLA signed email.

We have to wait for @bryk to get back as only he is able to actually configure it (if it's possible).

**Comment by floreks** [comment]
@nicr9 @joe2far I have 2 ideas :)

1. @nicr9 as you are the author of this PR then you could try to write "I authored these commits" and bot should approve it.

2. You can both try to write something like "I am okay with these commits beeing contributed to google".

Second option seems better as it would acknowledge both of you as authors. Let's see if it works.

**Comment by cheld** [comment]
Not sure if you changed something, but Github works ok now...

**Comment by nicr9** [comment]
I am okay with these commits being contributed to google

**Comment by nicr9** [comment]
I authored these commits

**Comment by joe2far** [comment]
I am okay with these commits beeing contributed to google

**Comment by floreks** [comment]
@googlebot check CLA

**Comment by floreks** [comment]
@joe2far unfortunately bot still doesn't accept your CLA. Just to be sure recheck that your default mail on GitHub, `git config user.email` and mail you have made your commits with are the same and have CLA signed.

**Comment by cheld** [comment]
I would like to test it. Do I have to deploy some helm containers to kubernetes? In the code I have seen a fixed IP address of a repository. Is it public available?

**Comment by joe2far** [comment]
@cheld - In order to test this, tiller needs to be installed in your k8s cluster, to do this: 
- first install the helm cli as from https://github.com/kubernetes/helm/blob/master/docs/install.md
- Then run the `helm init` command, this will install tiller in the k8s cluster configured in your ".kube" config.
- Then update your dashboard image to docker.io/ammeon/kubernetes-helm-dashboard-amd64:0.0.12` to test.

Or you can use `gulp serve` to see the GUI but this will not allow charts to be deployed as tiller is not exposed in the same when using `kubectl proxy`.

Yes, there is a hard coded chart repository that is internal / not public (it will just not show up in the available repos for you) and you will just see the k8s-charts repository as from https://github.com/kubernetes/charts .

The plan was to add a repository section to admin in the dashboard where additional chart repos could be added, which just the k8s-chart repo configured by default, but I wanted to get feedback before implementing this...

**Comment by nicr9** [comment]
@floreks Does googlebot rely only on the `user.email` in the commits or does it consider the `user.name` as well? Also: I believe @joe2far has changed his github handle recently but has updated the details in his CLA so that's probably not related.

**Comment by floreks** [comment]
@nicr9 I think it relies on `user.email` and tries to connect this information to user account on github.

**Comment by cheld** [comment]
BTW: you could make demo in sig ui meeting next wednesday

**Comment by joe2far** [comment]
@floreks - I have reviewed all the commits and they all contain my correct email address, which is as in my github account and with which, I have has signed the CLA... I am not sure why googlebot is still reporting a CLA problem?

**Comment by floreks** [comment]
@joe2far Ye, I thought your config is fine. The issue is then with the CLA bot. It has troubles identifying author of this PR. I'm not sure how it works exactly. Maybe all authors need to be assigned to PR and consent that they are ok with their commits being contributed. I can't assign you because you are not member of kubernetes organization.

Also I don't have access to configure anything so we have to wait for @bryk anyway. In the meantime we can review this. It is not so urgent to deal with CLA now.

**Comment by bryk** [comment]
I love this proof of concept! Thanks @nicr9 for putting this out. I think this is large enough topic that we should meet and talk about the progress here. How about we do a SIG-UI meeting dedicated to Helm support and go though all the questions there? We could also use it for scoping, figuring out the process and design.

**Comment by nicr9** [comment]
We're really glad you guys like it!

We'd be happy to meet and help flesh out a design but as there's still some open questions regarding helm's design plans maybe we should include the helm team or wait a while for helm to take shape.

**Comment by bryk** [comment]
> We'd be happy to meet and help flesh out a design but as there's still some open questions regarding helm's design plans maybe we should include the helm team or wait a while for helm to take shape.

Is this something you need to flesh out in the helm team, or can we help you? I'd really like to move this forward, so let us know if there is anything you need from us now.

**Comment by rf232** [comment]
Hello,

Is there an update on the status? I'd love to help you to get a good helm integration with the dashboard. I discussed with @bryk that I'll be the one responsible to help you guys land the feature and love to have a meeting with you to see what is needed from both sides to get this going.

**Comment by nicr9** [comment]
Hey @rf232,

Sorry about the radio silence lately! So far we don't have a lot to report; This was a much bigger feature than we initially anticipated. We're not 100% sure what kind of resources we could devote to this going forward and we're going to need to discuss this internally.

In the meantime, we'd love to meet and discuss the next steps. We have a short list for talking points:

* Other 3rd party dashboards/UIs want a REST API which Helm may expose in the future
* Is it okay to proceed with wrapping the Helm's gRPC API in the dashboard backend?
* UX design and prioritisation of features to expose

When would be a good time for a meeting?

**Comment by rf232** [comment]
Hi @nicr9 

I'm up for video call this week and the next week (except Wednesday) from 9 till 18 UTC. So pick your spot :)

**Comment by rf232** [comment]
https://docs.google.com/document/d/1U8lGzA-ykyDIJSaG7Hlc0y1kx8F78qrGmfUGuiKYlIY/edit?usp=sharing

I had a short discussion with Joe and Nic slightly over a week ago, here are the notes I took

Sorry it took me so long, but apparently I can't 'share' a doc I made on my corp account ;)

**Comment by gurvindersingh** [comment]
any update on this feature.

**Comment by nicr9** [comment]
Hey @gurvindersingh,

First off, for context; myself and @joe2far discussed this back in November with @rf232. We felt that, while this feature was definitely possible to achieve, it would involve a lot more work that we would be able to devote to it given our other commitments.

We put together a plan to divide the changes in this PR into (IIRC) 3 separate change sets that (individually) would be less ambitious but still be a step in the right direction. The intention was to shorten the gap needed to get the code to a standard that would be suitable for merging. When the change sets were available it may even be possible for someone else with more time to tackle them.

So you're probably wondering; where are those 3 change sets? Well the first step was to bring all dependencies for both projects (dashboard and helm) into line. Joe attempted this a while back (around the time that Helm 2.0 landed) and it's harder than it sounds! It will be a while before we get a chance to try again.

**tl;dr** - We still want to get this PR split up and moving forward but keeping up with ongoing dev in both projects is hard and this has become a spare-time endeavour for us.

**Comment by bryk** [comment]
Thanks for the update @nicr9!

**Comment by discostur** [comment]
Is there any update on this?

**Comment by floreks** [comment]
From our side it's currently on hold due to higher priority tasks. I think guys are also not working on it.

**Comment by maciaszczykm** [comment]
Closing as stale. Please reopen if you are willing to continue.
